components/collapsible-panel/collapsible-panel.js

```diff
@@ -452,21 +459,23 @@ class CollapsiblePanel extends SkeletonMixin(FocusMixin(RtlMixin(LitElement))) {
 	}
 
 	_renderPanelTitle() {
-		let headingStyle = (this.headingStyle === defaultHeading && this.headingLevel !== this.headingStyle) ? this.headingLevel : this.headingStyle;
-		headingStyle = normalizeHeadingNumber(headingStyle);
+		const headingStyle = (this.headingStyle === defaultHeading && this.headingLevel !== this.headingStyle) ? this.headingLevel : this.headingStyle;
+		this.headingStyle = normalizeHeadingStyle(headingStyle);
```

Comment:
I don't love mutating the property here, since if the normalization did something it's going to cause a re-render (and we're already in `render()`).

Reply:
Yeah that's a good point...I forgot why we did it this way.

That said, I'm not crazy about having an incorrect `heading-style` attribute on the DOM element...maybe the normalization would make more sense in `updated`? Then we just use `this.headingStyle` and `this.headingLevel` directly in this render method.

Reply:
I'm not too worried about it, but if you want to ensure that it's always valid I'd suggest creating a getter/setter and doing the validation in the setter. [Here's an example from color input](https://github.com/BrightspaceUI/core/blob/main/components/inputs/input-color.js#L248).

Reply:
Completely agree.

Also, for cases where we need to compute an internal reactive property in advance of `render()`, we have [willUpdate(...)](https://lit.dev/docs/components/lifecycle/#willupdate). It's rarely used in core, partly because a lot of the components were written before it existed. There's probably a few cases where we could avoid double rendering by using it. Anyway, just something to keep in mind.